src/NetAnalyzers/Core/Microsoft.CodeQuality.Analyzers/ApiDesignGuidelines/ConsiderPassingBaseTypesAsParameters.cs

```diff
@@ -0,0 +1,225 @@
+                        var parameterIndex = methodSymbol.GetParameterIndex(parameter);
+                        var matchingOverload = methodSymbol.GetMatchingOverload(overloads, parameterIndex, suggestedNamedType, context.CancellationToken);
+
+                        if (matchingOverload != null)
+                        {
+                            return;
+                        }
+                    }
+
+                    // All calls on this parameter are using the same base type which is different
+                    // from the one declared, so we can simply report.
+                    context.ReportDiagnostic(parameter.CreateDiagnostic(Rule, parameter.Name,
+                        parameter.Type.Name, suggestedBaseTypeSymbol.Name));
+                }
+            });
+        }
+
+        private static ITypeSymbol FindDeclaringType(ISymbol symbol)
+        {
+            if (symbol.IsOverride)
+            {
+                return FindDeclaringType(symbol.GetOverriddenMember());
+            }
+
+            var interfaceMembers = ExplicitOrImplicitInterfaceImplementations(symbol);
+            if (interfaceMembers.Length == 1)
+            {
+                return interfaceMembers[0].ContainingType;
+            }
+            else
+            {
+                // When reaching this point, either we found no interface member, which means the
+                // member is declared directly on the type OR more than one interface was found
+                // as declaring this type so let's assume it is defined on the containing type.
+                return symbol.ContainingType;
+            }
+        }
+
+        // Borrowed from Microsoft.CodeAnalysis.Workspaces/ISymbolExtensions.cs with a small tweak
+        // which consist in passing the member name.
+        private static ImmutableArray<ISymbol> ExplicitOrImplicitInterfaceImplementations(ISymbol symbol)
+        {
+            if (symbol.Kind != SymbolKind.Method &&
+                symbol.Kind != SymbolKind.Property &&
+                symbol.Kind != SymbolKind.Event)
+            {
+                return ImmutableArray<ISymbol>.Empty;
+            }
+
+            var containingType = symbol.ContainingType;
+            var query = from iface in containingType.AllInterfaces
+                        from interfaceMember in iface.GetMembers(symbol.Name)
+                        let impl = containingType.FindImplementationForInterfaceMember(interfaceMember)
+                        where symbol.Equals(impl)
+                        select interfaceMember;
+            return query.ToImmutableArray();
+        }
+
+        private static ITypeSymbol FindMostGenericType(ITypeSymbol originalType, ImmutableArray<ITypeSymbol> constraints)
+        {
+            Debug.Assert(constraints.Length >= 2);
+
+            var mostGenericInterface =
+                originalType.AllInterfaces.FirstOrDefault(type => constraints.All(x => type.Inherits(x)));
```

Comment:
Can we make this method more performant by ordering or reducing the `constraints` types set? For example, if it contains types A and B, such that A derives from B, then we can remove A from the constraints right?

Reply:
Is it going to be faster? I mean to make the reduction of the constraints I will have to make a lot of iterations.

Also, for your example, I think we want the opposite. If `A` derives from `B` then we want to keep `B` (i.e. the less generic type).

Reply:
Just in case, I was thinking about the following implementation:

```csharp
			var reducedConstraints = constraints.ToList();
            for (int i = 1; i < reducedConstraints.Count; i++)
            {
                if (reducedConstraints[i].Inherits(reducedConstraints[i - 1]))
                {
                    reducedConstraints.RemoveAt(i - 1);
                }
                else if (reducedConstraints[i - 1].Inherits(reducedConstraints[i]))
                {
                    reducedConstraints.RemoveAt(i);
                }
            }
```